### PR TITLE
[Feature] Add RotatedBoxes and QuadriBoxes data structure in mmrotate.

### PR DESCRIPTION
**Pull request by jbwang1997**
Thanks for your contribution and we appreciate it a lot. The following instructions would make your pull request more healthy and more easily get feedback. If you do not understand some items, don't worry, just make the pull request and seek help from maintainers.

## Motivation

Add RotatedBoxes and QuadriBoxes data structure in mmrotate.

- [x] Initialize RotatedBoxes and QuadriBoxes
- [x] Add converters for HorizontalBoxes, RotatedBoxes, and QuadriBoxes.
- [x] Add docstring for RotatedBoxes and QuadriBoxes
- [x] Add UT for all new features.

## Modification

Please briefly describe what modification is made in this PR.

## BC-breaking (Optional)

Does the modification introduce changes that break the back-compatibility of the downstream repos?
If so, please describe how it breaks the compatibility and how the downstream projects should modify their code to keep compatibility with this PR.

## Use cases (Optional)

If this PR introduces a new feature, it is better to list some use cases here, and update the documentation.

## Checklist

1. Pre-commit or other linting tools are used to fix the potential lint issues.
2. The modification is covered by complete unit tests. If not, please add more unit test to ensure the correctness.
3. The documentation has been modified accordingly, like docstring or example tutorials.

### CODE REVIEW

**Comment by zytx121** [comment]
In pipeline, bbox type is list[float]. If we use want to flip the rbboxes in pipeline, do we need to convert the list to Tensor firstly?

**Comment by jbwang1997** [comment]
All bboxes in the pipeline will be replaced by BoxList in the next step. `convert_bbox_mode` can also accept tensor and numpy data types.

File: mmrotate/core/bbox/structures/rotated_bbox.py [hunk 1]
```diff
@@ -0,0 +1,381 @@
...
+        """
+        bboxes = self.tensor
+        assert len(scale_factor) == 2
+        ctrs, wh, t = torch.split(bboxes, [2, 2, 1], dim=-1)
+        scale_factor = bboxes.new_tensor(scale_factor)
+        wh = wh * scale_factor
+        bboxes = torch.cat([ctrs, wh, t], dim=-1)
+        return type(self)(bboxes)
+
+    def is_bboxes_inside(self, img_shape: Tuple[int, int]) -> torch.BoolTensor:
```
Comment (BIGWangYuDong) [inline mmrotate/core/bbox/structures/rotated_bbox.py:334]: from torch import BoolTensor

File: mmrotate/core/bbox/structures/rotated_bbox.py [hunk 2]
```diff
@@ -0,0 +1,381 @@
...
+
+    def resize_bboxes(self: T, scale_factor: Tuple[float, float]) -> T:
+        """Resize the box width and height w.r.t scale_factor.
+
+        Args:
+            scale_factor (Tuple[float, float]): factors for scaling box
+                shapes. The length should be 2.
+
+        Returns:
+            Tensor: Resized bboxes with the same shape as the original boxes.
```
Comment (BIGWangYuDong) [inline mmrotate/core/bbox/structures/rotated_bbox.py:324]: T or Tensor

File: mmrotate/core/bbox/structures/bbox_mode_converters.py [hunk 3]
```diff
@@ -0,0 +1,112 @@
...
+from .quadrilateral_bbox import QuadriBoxes  # noqa
+from .rotated_bbox import RotatedBoxes  # noqa
+
+
+@register_bbox_mode_converter('hbox', 'rbox')
+def hbbox2rbbox(bboxes: Tensor) -> Tensor:
+    """Convert horizontal boxes to rotated boxes.
+
+    Args:
+        bboxes (Tensor): horizontal box tensor.
```
Comment (hhaAndroid) [inline mmrotate/core/bbox/structures/bbox_mode_converters.py:17]: Please specify the input and output shape

File: mmrotate/core/bbox/structures/bbox_mode_converters.py [hunk 4]
```diff
@@ -0,0 +1,112 @@
...
+    Returns:
+        Tensor: Rotated box tensor.
+    """
+    theta = bboxes.new_zeros((*bboxes.shape[:-1], 1))
+    return torch.cat([bboxes, theta], dim=-1)
+
+
+@register_bbox_mode_converter('hbox', 'qbox')
+def hbbox2qbbox(bboxes: Tensor) -> Tensor:
+    """Convert horizontal boxes to quadrilateral boxes.
```
Comment (hhaAndroid) [inline mmrotate/core/bbox/structures/bbox_mode_converters.py:28]: same as above

File: mmrotate/core/bbox/structures/bbox_mode_converters.py [hunk 5]
```diff
@@ -0,0 +1,112 @@
...
+    """Convert rotated boxes to horizontal boxes.
+
+    Args:
+        bboxes (Tensor): Rotated box tensor.
+
+    Returns:
+        Tensor: Horizontal box tensor.
+    """
+    ctrs, w, h, theta = torch.split(bboxes, (2, 1, 1, 1), dim=-1)
+    Cos, Sin = torch.cos(theta), torch.sin(theta)
```
Comment (hhaAndroid) [inline mmrotate/core/bbox/structures/bbox_mode_converters.py:51]: Why do you need to capitalize？

File: mmrotate/core/bbox/structures/quadrilateral_bbox.py [hunk 6]
```diff
@@ -0,0 +1,273 @@
...
+        return areas
+
+    @property
+    def widths(self) -> Tensor:
+        """Return a tensor representing the widths of boxes.
+
+        Quadrilateral boxes don't have the width concept. Use ``sqrt(areas)``
+        to replace the width.
+        """
+        return torch.sqrt(self.areas)
```
[the remark this answers is not in the document]
Comment (jbwang1997) [reply mmrotate/core/bbox/structures/quadrilateral_bbox.py:62]: the weights and heights are usually used to filter out invalid objects in pipelines. The influence is relatively small. I will add a warning here.

File: mmrotate/core/bbox/structures/bbox_mode_converters.py [hunk 7]
```diff
@@ -0,0 +1,114 @@
+# Copyright (c) OpenMMLab. All rights reserved.
+import cv2
+import numpy as np
+import torch
+from mmdet.structures.bbox import register_bbox_mode_converter
+from torch import Tensor
+
+from .quadrilateral_bbox import QuadriBoxes  # noqa
+from .rotated_bbox import RotatedBoxes  # noqa
```
[the remark this answers is not in the document]
Comment (jbwang1997) [reply mmrotate/core/bbox/structures/bbox_mode_converters.py:9]: QuadriBoxes and RotatedBoxes is not used in bbox_mode_converts. But, to ensure the box modes are registered before converters, I import them in bbox_mode_converters.

**Review by hhaAndroid — APPROVED** [review]

File: tests/test_core/test_bbox/test_structures/test_rotated_boxes.py [hunk 8]
```diff
@@ -0,0 +1,193 @@
...
+import cv2
+import numpy as np
+import torch
+from mmdet.structures.mask import BitmapMasks, PolygonMasks
+from mmengine.testing import assert_allclose
+
+from mmrotate.core.bbox.structures import RotatedBoxes
+
+
+class TestHorizontalBoxes(TestCase):
```
Comment (zytx121) [inline tests/test_core/test_bbox/test_structures/test_rotated_boxes.py:15]:
```suggestion
class TestRotatedBoxes(TestCase):
```

File: tests/test_core/test_bbox/test_structures/test_box_converters.py [hunk 9]
```diff
@@ -0,0 +1,78 @@
...
+from mmdet.structures.bbox import HorizontalBoxes
+from mmengine.testing import assert_allclose
+
+from mmrotate.core.bbox.structures import QuadriBoxes, RotatedBoxes
+
+
+class TestboxModeConverters(TestCase):
+
+    def test_hbox2rbox(self):
+        th_hboxes = torch.Tensor([10, 10, 20, 20]).reshape(1, 1, 4)
```
Comment (zytx121) [inline tests/test_core/test_bbox/test_structures/test_box_converters.py:16]: What's the meaning of the `th_` prefix?

Comment (jbwang1997) [reply tests/test_core/test_bbox/test_structures/test_box_converters.py:16]: th means torch. I want to represent it's a Tensor type boxes.

File: mmrotate/core/bbox/structures/box_converters.py [hunk 10]
```diff
@@ -0,0 +1,114 @@
...
+    """Convert quadrilateral boxes to rotated boxes.
+
+    Args:
+        boxes (Tensor): Quadrilateral box tensor with shape of (..., 8).
+
+    Returns:
+        Tensor: Rotated box tensor with shape of (..., 5).
+    """
+    original_shape = boxes.shape[:-1]
+    points = boxes.cpu().numpy().reshape(-1, 4, 2)
```
Comment (RangiLyu) [inline mmrotate/core/bbox/structures/box_converters.py:108]: May add a TODO here to support tensor-based minAreaRect later?

File: mmrotate/core/bbox/structures/quadri_boxes.py [hunk 11]
```diff
@@ -0,0 +1,337 @@
...
+
+    @property
+    def vertices(self) -> Tensor:
+        """Return a tensor representing the vertices of boxes."""
+        boxes = self.tensor
+        return boxes.reshape(*boxes.shape[:-1], 4, 2)
+
+    @property
+    def centers(self) -> Tensor:
+        """Return a tensor representing the centers of boxes."""
```
Comment (RangiLyu) [inline mmrotate/core/bbox/structures/quadri_boxes.py:48]: add tensor shape in the docstring

File: mmrotate/core/bbox/structures/quadri_boxes.py [hunk 12]
```diff
@@ -0,0 +1,337 @@
...
+        device (str or torch.device, Optional): device of boxes.
+            Default to None.
+        clone (bool): Whether clone ``boxes`` or not. Defaults to True.
+    """
+
+    box_dim = 8
+
+    @property
+    def vertices(self) -> Tensor:
+        """Return a tensor representing the vertices of boxes."""
```
Comment (RangiLyu) [inline mmrotate/core/bbox/structures/quadri_boxes.py:42]: add tensor shape in the docstring

**Review by RangiLyu — APPROVED** [review]

File: mmrotate/core/bbox/structures/quadri_boxes.py [hunk 13]
```diff
@@ -0,0 +1,359 @@
...
+
+        boxes = []
+        if isinstance(masks, BitmapMasks):
+            for idx in range(num_masks):
+                mask = masks.masks[idx]
+                points = np.stack(np.nonzero(mask), axis=-1).astype(np.float32)
+                rect = cv2.minAreaRect(points)
+                (x1, y1), (x2, y2), (x3, y3), (x4, y4) = cv2.boxPoints(rect)
+                boxes.append([x1, y1, x2, y2, x3, y3, x4, y4])
+        elif isinstance(masks, PolygonMasks):
```
Comment (RangiLyu) [inline mmrotate/core/bbox/structures/quadri_boxes.py:345]: The BaseInstanceMasks has to_ndarray method, which can unify this part.

**Review by RangiLyu — APPROVED** [review]

**Pull request merged**